zerver/tornado/event_queue.py

```diff
@@ -1622,6 +1622,21 @@ def process_user_group_name_update_event(event: Mapping[str, Any], users: Iterab
                 client.add_event(user_group_event)
 
 
+def process_user_topic_event(event: Mapping[str, Any], users: Iterable[int]) -> None:
+    for user_profile_id in users:
+        for client in get_client_descriptors_for_user(user_profile_id):
+            if not client.accepts_event(event):
+                continue
+
+            if event.get("topic_name") != "" or client.empty_topic_name:
+                client.add_event(event)
+                continue
+
+            event_copy = dict(event)
+            event_copy["topic_name"] = Message.EMPTY_TOPIC_FALLBACK_NAME
+            client.add_event(event_copy)
```

Comment:
The optimization that would work here is something like:
```
if event.get("topic_name") == "":
    empty_topic_name_fallback_event = dict(event)
    empty_topic_name_fallback_event["topic_name"] = Message.EMPTY_TOPIC_FALLBACK_NAME
else: 
    empty_topic_name_fallback_event = event
```

and then inside the loop, just check `client.empty_topic_name` and do `add_event` on either `event` or `empty_topic_name_fallback_event`.

Then we only construct the fallback event once per event, not once per client without the client capability (which will be all of them).

Reply:
Made the changes